graphql_compiler/compiler/ir_lowering_sql/__init__.py

```diff
@@ -1,35 +1,148 @@
-                                *****
-                                Be very careful with this option, as bad input here will
-                                lead to incorrect output queries being generated.
-                                *****
+        type_equivalence_hints: optional dict of GraphQL interface or type -> GraphQL union. Unused.
 
     Returns:
         tree representation of IR blocks for recursive traversal by SQL backend.
     """
-    raise NotImplementedError(u'SQL IR lowering is not yet implemented.')
+    construct_result = ir_blocks.pop()
+    query_path_to_location_info = _map_query_path_to_location_info(query_metadata_table)
+    query_path_to_output_fields = _map_query_path_to_outputs(
+        construct_result, query_path_to_location_info)
+
+    # iteratively construct SqlTree
+    query_path_to_node = {}
+    block_index_to_location = _map_block_index_to_location(ir_blocks)
+    tree_root = None
+    for index, block in enumerate(ir_blocks):
+        location = block_index_to_location[index]
+        if isinstance(block, (blocks.QueryRoot,)):
+            query_path = location.query_path
+            if tree_root is not None:
+                raise AssertionError(
+                    u'Encountered QueryRoot {} but tree root is already set to {} during '
+                    u'construction of SQL query tree for IR blocks {} with query'
+                    u' metadata table {}'.format(
+                        block, tree_root, ir_blocks, query_metadata_table))
+            tree_root = SqlNode(block=block, query_path=query_path)
+            query_path_to_node[query_path] = tree_root
+        elif isinstance(block, _SKIPPABLE_BLOCKS):
+            continue
+        else:
+            raise exceptions.GraphQLNotSupportedByBackendError(
+                u'Encountered unsupported block {} during construction of SQL query tree for IR '
+                u'blocks {} with query metadata table {} .'.format(
+                    block, ir_blocks, query_metadata_table))
+
+    return SqlQueryTree(tree_root, query_path_to_location_info, query_path_to_output_fields)
+
+
+def _map_query_path_to_location_info(query_metadata_table):
+    """Create a map from a query path to a LocationInfo at that path.
+
+    Verifies that LocationInfos at the same query path are equivalent in type, parent,
+    recursive_depth and optional_scopes_depth.
+
+    Args:
+        query_metadata_table: QueryMetadataTable object containing all metadata collected during
+                              query processing, including location metadata (e.g. which locations
+                              are folded or optional).
+
+    Returns:
+        Dict[Tuple[str], LocationInfo], dictionary mapping query path to LocationInfo at that path.
+    """
+    query_path_to_location_info = {}
+    for location, location_info in query_metadata_table.registered_locations:
+        if location.query_path in query_path_to_location_info:
```

Comment:
Not all `BaseLocation` objects (the type inside the iterable at `query_metadata_table.registered_locations`) have a `query_path` attribute. You need to filter out the ones that are instances of `Location` here.

Reply:
I didn't realize that, when does this occur?

Reply:
Locations inside a `@fold` are of type `FoldScopeLocation` which does not have a `query_path`. In principle, there may be more examples in the future.